src/main/java/org/usfirst/frc4048/common/swervev3/estimation/PoseEstimator.java

```diff
@@ -0,0 +1,141 @@
+    private static final Vector<N3> visionMeasurementStdDevs1 = VecBuilder.fill(0.5, 0.5, 0.5);
+
+    /* standard deviation of vision readings, the lower the numbers arm, the more we trust vision */
+    private static final Vector<N3> visionMeasurementStdDevs2 = VecBuilder.fill(0.45, 0.45, 0.001);
+    private static final Transform2d cameraOneTransform = new Transform2d(Constants.CAMERA_OFFSET_FROM_CENTER_X, Constants.CAMERA_OFFSET_FROM_CENTER_Y, new Rotation2d());
+    private static final Transform2d cameraTwoTransform = new Transform2d(Constants.CAMERA_OFFSET_FROM_CENTER_X, Constants.CAMERA_OFFSET_FROM_CENTER_Y, new Rotation2d());
+    private final PoseManager poseManager;
+
+    public PoseEstimator(Module frontLeftMotor, Module frontRightMotor, Module backLeftMotor, Module backRightMotor, ApriltagIO apriltagIO, SwerveDriveKinematics kinematics, double initGyroValueDeg) {
+        this.frontLeft = frontLeftMotor;
+        this.frontRight = frontRightMotor;
+        this.backLeft = backLeftMotor;
+        this.backRight = backRightMotor;
+        this.apriltagSystem = new LoggableSystem<>(apriltagIO, new ApriltagInputs());
+        OdometryMeasurement initMeasurement = new OdometryMeasurement(
+                new SwerveModulePosition[]{
+                        frontLeft.getPosition(),
+                        frontRight.getPosition(),
+                        backLeft.getPosition(),
+                        backRight.getPosition(),
+                }, initGyroValueDeg
+        );
+        TimeInterpolatableBuffer<Pose2d> m1Buffer = TimeInterpolatableBuffer.createBuffer(Constants.POSE_BUFFER_STORAGE_TIME);
+        this.poseManager = new FilterablePoseManager(stateStdDevs1, visionMeasurementStdDevs1, kinematics, initMeasurement, m1Buffer, new BasicVisionFilter(m1Buffer){
+            @Override
+            public Pose2d getVisionPose(VisionMeasurement measurement) {
+                return measurement.measurement().plus(cameraOneTransform);
+            }
+        });
+        SmartDashboard.putData(field);
+    }
+
+    public void updateInputs() {
+        apriltagSystem.updateInputs();
+    }
+
+    /**
+     * updates odometry, should be called in periodic
+     *
+     * @see SwerveDrivePoseEstimator#update(Rotation2d, SwerveModulePosition[])
+     */
+    public void updatePosition(OdometryMeasurement m) {
+        if (DriverStation.isEnabled()) {
+            poseManager.addOdomMeasurement(m, Logger.getRealTimestamp());
+        }
+        field.setRobotPose(poseManager.getEstimatedPosition());
+    }
+
+    private boolean validAprilTagPose(double[] measurement) {
+        return !ArrayUtils.allMatch(measurement, -1.0) && measurement.length == 3;
+    }
+
+    public void updateVision() {
+        if (ExampleAdvantageScopeRobot.getMode().equals(RobotMode.TELEOP) && Constants.ENABLE_VISION) {
+            for (int i = 0; i < apriltagSystem.getInputs().timestamp.length; i++) {
+                double[] pos = new double[]{
```

Comment:
added method docs. They are doubles because that is what the ApriltagIO is logging. It is in an array because there can be multiple apriltag measurements per tick.